File: platform-api/src/internal/rbac/permissions.go
/*
 *  Copyright (c) 2026, WSO2 LLC. (http://www.wso2.org) All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 */

package rbac

// Permission is a typed string identifying a single platform capability.
type Permission string

// Scope returns the OAuth2 scope string as Thunder issues it in the token (e.g. "ap:gateway:create").
// Thunder prefixes scopes with the resource-server identifier when issuing tokens.
func (p Permission) Scope() string {
	return "ap:" + string(p)
}

const (
	// Gateway
	GatewayManage       Permission = "gateway:manage"
	GatewayCreate       Permission = "gateway:create"
	GatewayRead         Permission = "gateway:read"
	GatewayUpdate       Permission = "gateway:update"
	GatewayDelete       Permission = "gateway:delete"
	GatewayTokenManage  Permission = "gateway:token:manage"
	GatewayTokenRead    Permission = "gateway:token:read"
	GatewayTokenCreate  Permission = "gateway:token:create"
	GatewayTokenDelete  Permission = "gateway:token:delete"
	GatewayPolicyManage Permission = "gateway:policy:manage"
	GatewayPolicyRead   Permission = "gateway:policy:read"
	GatewayPolicyCreate Permission = "gateway:policy:create"
	GatewayPolicyDelete Permission = "gateway:policy:delete"
	GatewayArtifactsRead Permission = "gateway:artifacts:read"
	GatewayManifestRead  Permission = "gateway:manifest:read"
	GatewayStatusRead    Permission = "gateway:status:read"

	// REST API
	APIManage  Permission = "rest_api:manage"
	APICreate  Permission = "rest_api:create"
	APIRead    Permission = "rest_api:read"
	APIUpdate  Permission = "rest_api:update"
	APIDelete  Permission = "rest_api:delete"
	APIPublish Permission = "rest_api:publish"
	APIImport  Permission = "rest_api:import"

	// REST API → Gateways (sub-resource scopes nested under rest_api)
	APIGatewayManage Permission = "rest_api:gateway:manage"
	APIGatewayCreate Permission = "rest_api:gateway:create"
	APIGatewayRead   Permission = "rest_api:gateway:read"

	// REST API → Deployments (sub-resource scopes nested under rest_api)
	APIDeploymentManage   Permission = "rest_api:deployment:manage"
	APIDeploymentCreate   Permission = "rest_api:deployment:create"
	APIDeploymentRead     Permission = "rest_api:deployment:read"
	APIDeploymentDelete   Permission = "rest_api:deployment:delete"
	APIDeploymentUndeploy Permission = "rest_api:deployment:undeploy"
	APIDeploymentRestore  Permission = "rest_api:deployment:restore"

	// Project
	ProjectManage Permission = "project:manage"
	ProjectCreate Permission = "project:create"
	ProjectRead   Permission = "project:read"
	ProjectUpdate Permission = "project:update"
	ProjectDelete Permission = "project:delete"

	// Application
	ApplicationManage    Permission = "application:manage"
	ApplicationCreate    Permission = "application:create"
	ApplicationRead      Permission = "application:read"
	ApplicationUpdate    Permission = "application:update"
	ApplicationDelete    Permission = "application:delete"

	// Application → API Keys (sub-resource)
	ApplicationAPIKeyManage Permission = "application:api_key:manage"
	ApplicationAPIKeyCreate Permission = "application:api_key:create"
	ApplicationAPIKeyRead   Permission = "application:api_key:read"
	ApplicationAPIKeyDelete Permission = "application:api_key:delete"

	// Application → Associations (sub-resource)
	ApplicationAssociationsManage    Permission = "application:associations:manage"
	ApplicationAssociationsCreate    Permission = "application:associations:create"
	ApplicationAssociationsRead      Permission = "application:associations:read"
	ApplicationAssociationsDelete    Permission = "application:associations:delete"
	ApplicationAssociationsAPIKeyRead Permission = "application:associations:api_key:read"

	// DevPortal — DevPortalManage is the root scope covering CRUD and lifecycle operations
	// (activate/deactivate/set-default).
	DevPortalManage Permission = "devportal:manage"
	DevPortalCreate Permission = "devportal:create"
	DevPortalRead   Permission = "devportal:read"
	DevPortalUpdate Permission = "devportal:update"
	DevPortalDelete Permission = "devportal:delete"

	// Subscription
	SubscriptionManage Permission = "subscription:manage"
	SubscriptionCreate Permission = "subscription:create"
	SubscriptionRead   Permission = "subscription:read"
	SubscriptionUpdate Permission = "subscription:update"
	SubscriptionDelete Permission = "subscription:delete"

	// SubscriptionPlan
	SubscriptionPlanManage Permission = "subscription_plan:manage"
	SubscriptionPlanCreate Permission = "subscription_plan:create"
	SubscriptionPlanRead   Permission = "subscription_plan:read"
	SubscriptionPlanUpdate Permission = "subscription_plan:update"
	SubscriptionPlanDelete Permission = "subscription_plan:delete"

	// REST API → API Keys (sub-resource scopes nested under rest_api)
	APIKeyManage Permission = "rest_api:api_key:manage"
	APIKeyCreate Permission = "rest_api:api_key:create"
	APIKeyRead   Permission = "rest_api:api_key:read"
	APIKeyUpdate Permission = "rest_api:api_key:update"
	APIKeyDelete Permission = "rest_api:api_key:delete"

	// LLM provider template
	LLMTemplateManage Permission = "llm_template:manage"
	LLMTemplateCreate Permission = "llm_template:create"
	LLMTemplateRead   Permission = "llm_template:read"
	LLMTemplateUpdate Permission = "llm_template:update"
	LLMTemplateDelete Permission = "llm_template:delete"

	// LLM provider
	LLMProviderManage             Permission = "llm_provider:manage"
	LLMProviderCreate             Permission = "llm_provider:create"
	LLMProviderRead               Permission = "llm_provider:read"
	LLMProviderUpdate             Permission = "llm_provider:update"
	LLMProviderDelete             Permission = "llm_provider:delete"
	LLMProviderDeploymentManage   Permission = "llm_provider:deployment:manage"
	LLMProviderDeploymentCreate   Permission = "llm_provider:deployment:create"
	LLMProviderDeploymentRead     Permission = "llm_provider:deployment:read"
	LLMProviderDeploymentDelete   Permission = "llm_provider:deployment:delete"
	LLMProviderDeploymentUndeploy Permission = "llm_provider:deployment:undeploy"
	LLMProviderDeploymentRestore  Permission = "llm_provider:deployment:restore"
	LLMProviderAPIKeyManage       Permission = "llm_provider:api_key:manage"
	LLMProviderAPIKeyCreate       Permission = "llm_provider:api_key:create"
	LLMProviderAPIKeyRead         Permission = "llm_provider:api_key:read"
	LLMProviderAPIKeyDelete       Permission = "llm_provider:api_key:delete"

	// LLM proxy
	LLMProxyManage             Permission = "llm_proxy:manage"
	LLMProxyCreate             Permission = "llm_proxy:create"
	LLMProxyRead               Permission = "llm_proxy:read"
	LLMProxyUpdate             Permission = "llm_proxy:update"
	LLMProxyDelete             Permission = "llm_proxy:delete"
	LLMProxyDeploymentManage   Permission = "llm_proxy:deployment:manage"
	LLMProxyDeploymentCreate   Permission = "llm_proxy:deployment:create"
	LLMProxyDeploymentRead     Permission = "llm_proxy:deployment:read"
	LLMProxyDeploymentDelete   Permission = "llm_proxy:deployment:delete"
	LLMProxyDeploymentUndeploy Permission = "llm_proxy:deployment:undeploy"
	LLMProxyDeploymentRestore  Permission = "llm_proxy:deployment:restore"
	LLMProxyAPIKeyManage       Permission = "llm_proxy:api_key:manage"
	LLMProxyAPIKeyCreate       Permission = "llm_proxy:api_key:create"
	LLMProxyAPIKeyRead         Permission = "llm_proxy:api_key:read"
	LLMProxyAPIKeyDelete       Permission = "llm_proxy:api_key:delete"

	// MCP proxy
	MCPProxyManage             Permission = "mcp_proxy:manage"
	MCPProxyCreate             Permission = "mcp_proxy:create"
	MCPProxyRead               Permission = "mcp_proxy:read"
	MCPProxyUpdate             Permission = "mcp_proxy:update"
	MCPProxyDelete             Permission = "mcp_proxy:delete"
	MCPProxyDeploymentManage   Permission = "mcp_proxy:deployment:manage"
	MCPProxyDeploymentCreate   Permission = "mcp_proxy:deployment:create"
	MCPProxyDeploymentRead     Permission = "mcp_proxy:deployment:read"
	MCPProxyDeploymentDelete   Permission = "mcp_proxy:deployment:delete"
	MCPProxyDeploymentUndeploy Permission = "mcp_proxy:deployment:undeploy"
	MCPProxyDeploymentRestore  Permission = "mcp_proxy:deployment:restore"

	// WebSub API
	WebSubAPIManage             Permission = "websub_api:manage"
	WebSubAPICreate             Permission = "websub_api:create"
	WebSubAPIRead               Permission = "websub_api:read"
	WebSubAPIUpdate             Permission = "websub_api:update"
	WebSubAPIDelete             Permission = "websub_api:delete"
	WebSubAPIDeploymentManage   Permission = "websub_api:deployment:manage"
	WebSubAPIDeploymentCreate   Permission = "websub_api:deployment:create"
	WebSubAPIDeploymentRead     Permission = "websub_api:deployment:read"
	WebSubAPIDeploymentDelete   Permission = "websub_api:deployment:delete"
	WebSubAPIDeploymentUndeploy Permission = "websub_api:deployment:undeploy"
	WebSubAPIDeploymentRestore  Permission = "websub_api:deployment:restore"
	WebSubAPIPublish            Permission = "websub_api:publish"
	WebSubAPIKeyManage          Permission = "websub_api:api_key:manage"
	WebSubAPIKeyCreate          Permission = "websub_api:api_key:create"
	WebSubAPIKeyUpdate          Permission = "websub_api:api_key:update"
	WebSubAPIKeyDelete          Permission = "websub_api:api_key:delete"

	// WebBroker API
	WebBrokerAPIManage             Permission = "webbroker_api:manage"
	WebBrokerAPICreate             Permission = "webbroker_api:create"
	WebBrokerAPIRead               Permission = "webbroker_api:read"
	WebBrokerAPIUpdate             Permission = "webbroker_api:update"
	WebBrokerAPIDelete             Permission = "webbroker_api:delete"
	WebBrokerAPIDeploymentManage   Permission = "webbroker_api:deployment:manage"
	WebBrokerAPIDeploymentCreate   Permission = "webbroker_api:deployment:create"
	WebBrokerAPIDeploymentRead     Permission = "webbroker_api:deployment:read"
	WebBrokerAPIDeploymentDelete   Permission = "webbroker_api:deployment:delete"
	WebBrokerAPIDeploymentUndeploy Permission = "webbroker_api:deployment:undeploy"
	WebBrokerAPIDeploymentRestore  Permission = "webbroker_api:deployment:restore"
	WebBrokerAPIPublish            Permission = "webbroker_api:publish"
	WebBrokerAPIKeyManage          Permission = "webbroker_api:api_key:manage"
	WebBrokerAPIKeyCreate          Permission = "webbroker_api:api_key:create"
	WebBrokerAPIKeyUpdate          Permission = "webbroker_api:api_key:update"
	WebBrokerAPIKeyDelete          Permission = "webbroker_api:api_key:delete"

	// Git
	GitRead Permission = "git:read"
)

// ManagedBy maps fine-grained write permissions to the resource-level manage permission
// that acts as a root scope covering all of them. A token carrying resource:manage is
// authorised for every permission listed here that maps to that manage scope.
// Read permissions and resource:manage itself are not included — they are checked directly.
var ManagedBy = map[Permission]Permission{
	// Gateway
	GatewayCreate:       GatewayManage,
	GatewayUpdate:       GatewayManage,
	GatewayDelete:       GatewayManage,
	GatewayTokenManage:  GatewayManage,
	GatewayTokenRead:    GatewayTokenManage,
	GatewayTokenCreate:  GatewayTokenManage,
	GatewayTokenDelete:  GatewayTokenManage,
	GatewayPolicyManage: GatewayManage,
	GatewayPolicyRead:   GatewayPolicyManage,
	GatewayPolicyCreate: GatewayPolicyManage,
	GatewayPolicyDelete: GatewayPolicyManage,
	GatewayArtifactsRead: GatewayManage,
	GatewayManifestRead:  GatewayManage,
	GatewayStatusRead:    GatewayManage,

	// API
	APICreate:  APIManage,
	APIUpdate:  APIManage,
	APIDelete:  APIManage,
	APIPublish: APIManage,
	APIImport:  APIManage,

	// API Gateways
	APIGatewayManage: APIManage,
	APIGatewayCreate: APIGatewayManage,

	// API Deployments
	APIDeploymentManage:   APIManage,
	APIDeploymentCreate:   APIDeploymentManage,
	APIDeploymentDelete:   APIDeploymentManage,
	APIDeploymentUndeploy: APIDeploymentManage,
	APIDeploymentRestore:  APIDeploymentManage,

	// Project
	ProjectCreate: ProjectManage,
	ProjectUpdate: ProjectManage,
	ProjectDelete: ProjectManage,

	// Application
	ApplicationCreate:    ApplicationManage,
	ApplicationUpdate:    ApplicationManage,
	ApplicationDelete:    ApplicationManage,
	ApplicationAPIKeyManage: ApplicationManage,
	ApplicationAPIKeyCreate: ApplicationAPIKeyManage,
	ApplicationAPIKeyDelete: ApplicationAPIKeyManage,
	ApplicationAssociationsManage:    ApplicationManage,
	ApplicationAssociationsCreate:    ApplicationAssociationsManage,
	ApplicationAssociationsDelete:    ApplicationAssociationsManage,

	// DevPortal
	DevPortalCreate: DevPortalManage,
	DevPortalUpdate: DevPortalManage,
	DevPortalDelete: DevPortalManage,

	// Subscription
	SubscriptionCreate: SubscriptionManage,
	SubscriptionUpdate: SubscriptionManage,
	SubscriptionDelete: SubscriptionManage,

	// SubscriptionPlan
	SubscriptionPlanCreate: SubscriptionPlanManage,
	SubscriptionPlanUpdate: SubscriptionPlanManage,
	SubscriptionPlanDelete: SubscriptionPlanManage,

	// APIKey
	APIKeyCreate: APIKeyManage,
	APIKeyUpdate: APIKeyManage,
	APIKeyDelete: APIKeyManage,

	// LLM provider template
	LLMTemplateCreate: LLMTemplateManage,
	LLMTemplateUpdate: LLMTemplateManage,
	LLMTemplateDelete: LLMTemplateManage,

	// LLM provider
	LLMProviderCreate:             LLMProviderManage,
	LLMProviderUpdate:             LLMProviderManage,
	LLMProviderDelete:             LLMProviderManage,
	LLMProviderDeploymentManage:   LLMProviderManage,
	LLMProviderDeploymentCreate:   LLMProviderDeploymentManage,
	LLMProviderDeploymentDelete:   LLMProviderDeploymentManage,
	LLMProviderDeploymentUndeploy: LLMProviderDeploymentManage,
	LLMProviderDeploymentRestore:  LLMProviderDeploymentManage,
	LLMProviderAPIKeyCreate:       LLMProviderAPIKeyManage,
	LLMProviderAPIKeyDelete:       LLMProviderAPIKeyManage,

	// LLM proxy
	LLMProxyCreate:             LLMProxyManage,
	LLMProxyUpdate:             LLMProxyManage,
	LLMProxyDelete:             LLMProxyManage,
	LLMProxyDeploymentManage:   LLMProxyManage,
	LLMProxyDeploymentCreate:   LLMProxyDeploymentManage,
	LLMProxyDeploymentDelete:   LLMProxyDeploymentManage,
	LLMProxyDeploymentUndeploy: LLMProxyDeploymentManage,
	LLMProxyDeploymentRestore:  LLMProxyDeploymentManage,
	LLMProxyAPIKeyCreate:       LLMProxyAPIKeyManage,
	LLMProxyAPIKeyDelete:       LLMProxyAPIKeyManage,

	// MCP proxy
	MCPProxyCreate:             MCPProxyManage,
	MCPProxyUpdate:             MCPProxyManage,
	MCPProxyDelete:             MCPProxyManage,
	MCPProxyDeploymentManage:   MCPProxyManage,
	MCPProxyDeploymentCreate:   MCPProxyDeploymentManage,
	MCPProxyDeploymentDelete:   MCPProxyDeploymentManage,
	MCPProxyDeploymentUndeploy: MCPProxyDeploymentManage,
	MCPProxyDeploymentRestore:  MCPProxyDeploymentManage,

	// WebSub API
	WebSubAPICreate:             WebSubAPIManage,
	WebSubAPIUpdate:             WebSubAPIManage,
	WebSubAPIDelete:             WebSubAPIManage,
	WebSubAPIDeploymentManage:   WebSubAPIManage,
	WebSubAPIDeploymentCreate:   WebSubAPIDeploymentManage,
	WebSubAPIDeploymentDelete:   WebSubAPIDeploymentManage,
	WebSubAPIDeploymentUndeploy: WebSubAPIDeploymentManage,
	WebSubAPIDeploymentRestore:  WebSubAPIDeploymentManage,
	WebSubAPIPublish:            WebSubAPIManage,
	WebSubAPIKeyCreate:          WebSubAPIKeyManage,
	WebSubAPIKeyUpdate:          WebSubAPIKeyManage,
	WebSubAPIKeyDelete:          WebSubAPIKeyManage,

	// WebBroker API
	WebBrokerAPICreate:             WebBrokerAPIManage,
	WebBrokerAPIUpdate:             WebBrokerAPIManage,
	WebBrokerAPIDelete:             WebBrokerAPIManage,
	WebBrokerAPIDeploymentManage:   WebBrokerAPIManage,
	WebBrokerAPIDeploymentCreate:   WebBrokerAPIDeploymentManage,
	WebBrokerAPIDeploymentDelete:   WebBrokerAPIDeploymentManage,
	WebBrokerAPIDeploymentUndeploy: WebBrokerAPIDeploymentManage,
	WebBrokerAPIDeploymentRestore:  WebBrokerAPIDeploymentManage,
	WebBrokerAPIPublish:            WebBrokerAPIManage,
	WebBrokerAPIKeyCreate:          WebBrokerAPIKeyManage,
	WebBrokerAPIKeyUpdate:          WebBrokerAPIKeyManage,
	WebBrokerAPIKeyDelete:          WebBrokerAPIKeyManage,
}

File: portals/ai-workspace/src/auth/permissions.ts
/*
 * Copyright (c) 2026, WSO2 LLC. (https://www.wso2.com).
 *
 * WSO2 LLC. licenses this file to you under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

export type PlatformRole = 'admin' | 'developer' | 'publisher' | 'operator' | 'viewer';

export const PLATFORM_ROLES: readonly PlatformRole[] = ['admin', 'developer', 'publisher', 'operator', 'viewer'];

export function isPlatformRole(value: unknown): value is PlatformRole {
  return typeof value === 'string' && (PLATFORM_ROLES as readonly string[]).includes(value);
}

/**
 * Tooltip shown on a control disabled purely because the signed-in user lacks
 * the required scope. Deliberately does not name the missing scope — that would
 * let a caller probe the authorization model.
 */
export const NO_PERMISSION_TOOLTIP =
  'You do not have permission to perform this action. Please contact your admin.';

/**
 * `sx` for a permission-disabled action that renders as a link
 * (`component={RouterLink}`), i.e. an `<a>` rather than a `<button>`.
 *
 * A disabled `<a>` still blocks the click, but it does not pick up the
 * theme's disabled colouring the way a native disabled `<button>` does, so the
 * control looks fully enabled while doing nothing — which reads as a broken
 * button rather than a permission boundary. Dimming it makes the disabled state
 * visible. Not needed for plain `<Button onClick=...>`.
 */
export const DISABLED_ACTION_SX = {
  '&.Mui-disabled': { opacity: 0.55 },
} as const;

/** All platform API OAuth2 scopes derived from openapi.yaml x-required-scopes (ap: prefix). */
export const SCOPES = {
  // Organization
  ORGANIZATION_READ:              'ap:organization:read',
  ORGANIZATION_MANAGE:            'ap:organization:manage',

  // Projects
  PROJECT_READ:   'ap:project:read',
  PROJECT_CREATE: 'ap:project:create',
  PROJECT_UPDATE: 'ap:project:update',
  PROJECT_DELETE: 'ap:project:delete',
  PROJECT_MANAGE: 'ap:project:manage',

  // Applications
  APPLICATION_READ:                    'ap:application:read',
  APPLICATION_CREATE:                  'ap:application:create',
  APPLICATION_UPDATE:                  'ap:application:update',
  APPLICATION_DELETE:                  'ap:application:delete',
  APPLICATION_MANAGE:                  'ap:application:manage',
  APPLICATION_API_KEY_READ:            'ap:application:api_key:read',
  APPLICATION_API_KEY_CREATE:          'ap:application:api_key:create',
  APPLICATION_API_KEY_DELETE:          'ap:application:api_key:delete',
  APPLICATION_API_KEY_MANAGE:          'ap:application:api_key:manage',
  APPLICATION_ASSOCIATIONS_READ:       'ap:application:association:read',
  APPLICATION_ASSOCIATIONS_CREATE:     'ap:application:association:create',
  APPLICATION_ASSOCIATIONS_DELETE:     'ap:application:association:delete',
  APPLICATION_ASSOCIATIONS_MANAGE:     'ap:application:association:manage',
  APPLICATION_ASSOCIATIONS_API_KEY_READ: 'ap:application:association:api_key:read',

  // AI Gateways
  GATEWAY_READ:           'ap:gateway:read',
  GATEWAY_CREATE:         'ap:gateway:create',
  GATEWAY_UPDATE:         'ap:gateway:update',
  GATEWAY_DELETE:         'ap:gateway:delete',
  GATEWAY_MANAGE:         'ap:gateway:manage',
  GATEWAY_TOKEN_READ:     'ap:gateway:token:read',
  GATEWAY_TOKEN_CREATE:   'ap:gateway:token:create',
  GATEWAY_TOKEN_DELETE:   'ap:gateway:token:delete',
  GATEWAY_TOKEN_MANAGE:   'ap:gateway:token:manage',
  GATEWAY_CUSTOM_POLICY_READ:    'ap:gateway_custom_policy:read',
  GATEWAY_CUSTOM_POLICY_CREATE:  'ap:gateway_custom_policy:create',
  GATEWAY_CUSTOM_POLICY_DELETE:  'ap:gateway_custom_policy:delete',
  GATEWAY_CUSTOM_POLICY_MANAGE:  'ap:gateway_custom_policy:manage',
  GATEWAY_ARTIFACTS_READ: 'ap:gateway:artifact:read',
  GATEWAY_MANIFEST_READ:  'ap:gateway:manifest:read',

  // REST APIs
  REST_API_READ:              'ap:rest_api:read',
  REST_API_CREATE:            'ap:rest_api:create',
  REST_API_UPDATE:            'ap:rest_api:update',
  REST_API_DELETE:            'ap:rest_api:delete',
  REST_API_MANAGE:            'ap:rest_api:manage',
  REST_API_GATEWAY_READ:      'ap:rest_api:gateway:read',
  REST_API_GATEWAY_CREATE:    'ap:rest_api:gateway:create',
  REST_API_GATEWAY_MANAGE:    'ap:rest_api:gateway:manage',
  REST_API_DEPLOYMENT_READ:   'ap:rest_api:deployment:read',
  REST_API_DEPLOYMENT_CREATE: 'ap:rest_api:deployment:create',
  REST_API_DEPLOYMENT_DELETE: 'ap:rest_api:deployment:delete',
  REST_API_DEPLOYMENT_MANAGE: 'ap:rest_api:deployment:manage',
  REST_API_API_KEY_READ:      'ap:rest_api:api_key:read',
  REST_API_API_KEY_CREATE:    'ap:rest_api:api_key:create',
  REST_API_API_KEY_UPDATE:    'ap:rest_api:api_key:update',
  REST_API_API_KEY_DELETE:    'ap:rest_api:api_key:delete',
  REST_API_API_KEY_MANAGE:    'ap:rest_api:api_key:manage',
  REST_API_PUBLICATION_READ:  'ap:rest_api:publication:read',

  // Subscriptions
  SUBSCRIPTION_READ:        'ap:subscription:read',
  SUBSCRIPTION_CREATE:      'ap:subscription:create',
  SUBSCRIPTION_UPDATE:      'ap:subscription:update',
  SUBSCRIPTION_DELETE:      'ap:subscription:delete',
  SUBSCRIPTION_MANAGE:      'ap:subscription:manage',
  SUBSCRIPTION_PLAN_READ:   'ap:subscription_plan:read',
  SUBSCRIPTION_PLAN_CREATE: 'ap:subscription_plan:create',
  SUBSCRIPTION_PLAN_UPDATE: 'ap:subscription_plan:update',
  SUBSCRIPTION_PLAN_DELETE: 'ap:subscription_plan:delete',
  SUBSCRIPTION_PLAN_MANAGE: 'ap:subscription_plan:manage',

  // LLM Templates
  LLM_TEMPLATE_READ:   'ap:llm_template:read',
  LLM_TEMPLATE_CREATE: 'ap:llm_template:create',
  LLM_TEMPLATE_UPDATE: 'ap:llm_template:update',
  LLM_TEMPLATE_DELETE: 'ap:llm_template:delete',
  LLM_TEMPLATE_MANAGE: 'ap:llm_template:manage',

  // LLM Providers
  LLM_PROVIDER_READ:              'ap:llm_provider:read',
  LLM_PROVIDER_CREATE:            'ap:llm_provider:create',
  LLM_PROVIDER_UPDATE:            'ap:llm_provider:update',
  LLM_PROVIDER_DELETE:            'ap:llm_provider:delete',
  LLM_PROVIDER_MANAGE:            'ap:llm_provider:manage',
  LLM_PROVIDER_API_KEY_READ:      'ap:llm_provider:api_key:read',
  LLM_PROVIDER_API_KEY_CREATE:    'ap:llm_provider:api_key:create',
  LLM_PROVIDER_API_KEY_DELETE:    'ap:llm_provider:api_key:delete',
  LLM_PROVIDER_API_KEY_MANAGE:    'ap:llm_provider:api_key:manage',
  LLM_PROVIDER_DEPLOYMENT_READ:   'ap:llm_provider:deployment:read',
  LLM_PROVIDER_DEPLOYMENT_CREATE: 'ap:llm_provider:deployment:create',
  LLM_PROVIDER_DEPLOYMENT_DELETE: 'ap:llm_provider:deployment:delete',
  LLM_PROVIDER_DEPLOYMENT_MANAGE: 'ap:llm_provider:deployment:manage',

  // LLM Proxies
  LLM_PROXY_READ:              'ap:llm_proxy:read',
  LLM_PROXY_CREATE:            'ap:llm_proxy:create',
  LLM_PROXY_UPDATE:            'ap:llm_proxy:update',
  LLM_PROXY_DELETE:            'ap:llm_proxy:delete',
  LLM_PROXY_MANAGE:            'ap:llm_proxy:manage',
  LLM_PROXY_API_KEY_READ:      'ap:llm_proxy:api_key:read',
  LLM_PROXY_API_KEY_CREATE:    'ap:llm_proxy:api_key:create',
  LLM_PROXY_API_KEY_DELETE:    'ap:llm_proxy:api_key:delete',
  LLM_PROXY_API_KEY_MANAGE:    'ap:llm_proxy:api_key:manage',
  LLM_PROXY_DEPLOYMENT_READ:   'ap:llm_proxy:deployment:read',
  LLM_PROXY_DEPLOYMENT_CREATE: 'ap:llm_proxy:deployment:create',
  LLM_PROXY_DEPLOYMENT_DELETE: 'ap:llm_proxy:deployment:delete',
  LLM_PROXY_DEPLOYMENT_MANAGE: 'ap:llm_proxy:deployment:manage',

  // MCP Proxies
  MCP_PROXY_READ:              'ap:mcp_proxy:read',
  MCP_PROXY_CREATE:            'ap:mcp_proxy:create',
  MCP_PROXY_UPDATE:            'ap:mcp_proxy:update',
  MCP_PROXY_DELETE:            'ap:mcp_proxy:delete',
  MCP_PROXY_MANAGE:            'ap:mcp_proxy:manage',
  MCP_PROXY_DEPLOYMENT_READ:   'ap:mcp_proxy:deployment:read',
  MCP_PROXY_DEPLOYMENT_CREATE: 'ap:mcp_proxy:deployment:create',
  MCP_PROXY_DEPLOYMENT_DELETE: 'ap:mcp_proxy:deployment:delete',
  MCP_PROXY_DEPLOYMENT_MANAGE: 'ap:mcp_proxy:deployment:manage',

  // WebSub APIs
  WEBSUB_API_READ:              'ap:websub_api:read',
  WEBSUB_API_CREATE:            'ap:websub_api:create',
  WEBSUB_API_UPDATE:            'ap:websub_api:update',
  WEBSUB_API_DELETE:            'ap:websub_api:delete',
  WEBSUB_API_MANAGE:            'ap:websub_api:manage',
  WEBSUB_API_API_KEY_MANAGE:    'ap:websub_api:api_key:manage',
  WEBSUB_API_DEPLOYMENT_MANAGE: 'ap:websub_api:deployment:manage',
  WEBSUB_API_PUBLICATION_READ:  'ap:websub_api:publication:read',

  // WebBroker APIs
  WEBBROKER_API_READ:              'ap:webbroker_api:read',
  WEBBROKER_API_CREATE:            'ap:webbroker_api:create',
  WEBBROKER_API_UPDATE:            'ap:webbroker_api:update',
  WEBBROKER_API_DELETE:            'ap:webbroker_api:delete',
  WEBBROKER_API_MANAGE:            'ap:webbroker_api:manage',
  WEBBROKER_API_API_KEY_MANAGE:    'ap:webbroker_api:api_key:manage',
  WEBBROKER_API_DEPLOYMENT_MANAGE: 'ap:webbroker_api:deployment:manage',
  WEBBROKER_API_PUBLICATION_READ:  'ap:webbroker_api:publication:read',

  // Secrets
  SECRET_READ:   'ap:secret:read',
  SECRET_CREATE: 'ap:secret:create',
  SECRET_UPDATE: 'ap:secret:update',
  SECRET_DELETE: 'ap:secret:delete',
  SECRET_MANAGE: 'ap:secret:manage',
} as const;

/**
 * Deployment scopes per deployable AI artifact kind, keyed by the same resource
 * type `GatewayDeployProvider` takes. Shared by the deploy page/context and the
 * "Deploy to Gateway" entry points on the overview pages, so one map decides
 * both whether the deploy actions render and whether the button leading to them
 * is reachable.
 */
export const DEPLOYMENT_SCOPES = {
  provider: {
    read: SCOPES.LLM_PROVIDER_DEPLOYMENT_READ,
    create: SCOPES.LLM_PROVIDER_DEPLOYMENT_CREATE,
  },
  proxy: {
    read: SCOPES.LLM_PROXY_DEPLOYMENT_READ,
    create: SCOPES.LLM_PROXY_DEPLOYMENT_CREATE,
  },
  'mcp-server': {
    read: SCOPES.MCP_PROXY_DEPLOYMENT_READ,
    create: SCOPES.MCP_PROXY_DEPLOYMENT_CREATE,
  },
} as const;

export type DeployableResourceType = keyof typeof DEPLOYMENT_SCOPES;

/**
 * Scopes that must be held explicitly and are never derived from a broader
 * `:manage`. `ap:api_key:all:manage` is an ownership override (it widens which
 * users' keys are reachable, not which actions), so holding `ap:api_key:manage`
 * must not confer it — matching the service-layer rule in
 * `platform-api/internal/service` and `/me/api-keys` in openapi.yaml, the one
 * operation whose accepted-scope list omits its parent `:manage`.
 */
const NON_DERIVABLE_SCOPE_SUFFIX = ':all:manage';

/**
 * Check whether a set of scopes grants a requested scope.
 *
 * Mirrors how platform-api actually authorizes a request: each operation in
 * openapi.yaml declares a list of accepted scopes, and `ScopeEnforcer`
 * (`platform-api/internal/middleware/authorization.go`) admits the caller if a
 * held scope matches an entry in that list exactly — there is no wildcard form.
 * Rules:
 *
 *  1. Exact match — the scope is directly present.
 *  2. Own-level `:manage` — an operation accepting `<level>:<action>` also
 *     accepts `<level>:manage` (`ap:llm_provider:api_key:read` is satisfied by
 *     `ap:llm_provider:api_key:manage`).
 *  3. Ancestor `:manage` — a resource-level `:manage` also covers its
 *     sub-resources' operations, because those operations list it explicitly.
 *     `ap:llm_provider:manage` appears in the accepted list of every
 *     `/llm-providers/{id}/api-keys` and `/deployments` operation, so it grants
 *     `ap:llm_provider:api_key:create`, `ap:llm_provider:deployment:read`, etc.
 *     This holds for 47 of the 48 sub-resource operations in the spec; the sole
 *     exception is the override scope excluded above.
 */
export function checkPermission(userScopes: string[], scope: string): boolean {
  if (userScopes.includes(scope)) return true;

  const parts = scope.split(':');
  if (parts.length < 2) return false;

  if (scope.endsWith(NON_DERIVABLE_SCOPE_SUFFIX)) return false;

  // Own-level `:manage`, then each broader ancestor's `:manage`. For
  // `ap:llm_provider:api_key:read` that is `ap:llm_provider:api_key:manage`
  // followed by `ap:llm_provider:manage`.
  for (let depth = parts.length - 1; depth >= 2; depth -= 1) {
    const candidate = `${parts.slice(0, depth).join(':')}:manage`;
    if (candidate !== scope && userScopes.includes(candidate)) return true;
  }

  return false;
}
